onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2832,447 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .TypeConstraint("U", {"tensor(float)"}, "Constrain mean and inv_std_var to float tensors.")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          propagateShapeAndTypeFromFirstInput(ctx);
+          auto stash_type = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+          if (ctx.getNumOutputs() > 1) {
+            auto output_type = ctx.getOutputType(1);
+            output_type->mutable_tensor_type()->set_elem_type(static_cast<int32_t>(stash_type));
+          }
+          if (ctx.getNumOutputs() > 2) {
+            auto output_type = ctx.getOutputType(2);
+            output_type->mutable_tensor_type()->set_elem_type(static_cast<int32_t>(stash_type));
+          }
+          if (ctx.getNumOutputs() > 3) {
+            propagateElemTypeFromInputToOutput(ctx, 0, 3);
+          }
+          if (!hasNInputShapes(ctx, 1)) {
+            return;
+          }
+          auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+          int64_t input_ndim = input_shape.dim_size();
+          int axis = static_cast<int>(input_ndim - 1);
+
+          if (ctx.getNumOutputs() > 1) {
+            auto mean_shape = ctx.getOutputType(1)->mutable_tensor_type()->mutable_shape();
+            mean_shape->CopyFrom(input_shape);
+            mean_shape->mutable_dim(axis)->set_dim_value(1);
+          }
+
+          if (ctx.getNumOutputs() > 2) {
+            auto inv_std_dev_shape = ctx.getOutputType(2)->mutable_tensor_type()->mutable_shape();
+            inv_std_dev_shape->CopyFrom(input_shape);
+            inv_std_dev_shape->mutable_dim(axis)->set_dim_value(1);
+          }
+
+          if (ctx.getNumOutputs() > 3) {
+            propagateShapeFromInputToOutput(ctx, 0, 3);
+          }
+        })
+        .SetContextDependentFunctionBodyBuilder(
+            [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+              // SkipLayerNormalization  <epsilon> (X, S, gamma, beta, B) => (Y, Mean?, InvStdDev?, InputSkipBiasSum)
+
+              int64_t U = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+              auto* epsilon_attr = ctx.getAttribute("epsilon");
+              float epsilon = (epsilon_attr != nullptr) ? epsilon_attr->f() : 1e-5f;
+
+              FunctionBuilder builder(functionProto);
+              builder.Const("FloatEpsilon", ToTensor<float>(epsilon))
+                .Add("Epsilon = Cast (FloatEpsilon)", "to", U);
+              // Check if bias needs to be added to the sum
+              if (ctx.hasInput(4)) {
+                builder.Add("InputSkipSum = Add (X, S)")
+                    .Add("LNInput = Add (InputSkipSum, B)");
+              } else {
+                builder.Add("LNInput = Add (X, S)");
+              }
+              // Check if beta is an input to LayerNormalization
+              if (ctx.hasInput(3)) {
+                builder.Add("Y = LayerNormalization <axes = -1> (X, gamma, beta)", "epsilon", epsilon);
```

Comment:
Is the input of layer-norm supposed to be `LNInput` or `X` ?

Reply:
LNInput